### PR TITLE
Use

### PR DESCRIPTION
**Pull request by ola-rozenfeld**
Also adding max concurrency computation in the CAS fake and checking it is as expected.

### CODE REVIEW

**Comment by ola-rozenfeld** [comment]
Hmm. I had a commit description but git eated it :-)
This uses Chunkers in uploads. I still keep the original API of using byte[] and map[Digest][]byte for now, mostly because I don't want to break existing users.

File: go/client/cas.go [hunk 1]
```diff
@@ -11,20 +11,20 @@ import (
...
+	log "github.com/golang/glog"
 )
 
-// WriteBlobs stores a large number of blobs from a digest-to-blob map. It's intended for use on the
-// result of PackageTree. Unlike with the single-item functions, it first queries the CAS to
-// see which blobs are missing and only uploads those that are.
-func (c *Client) WriteBlobs(ctx context.Context, blobs map[digest.Digest][]byte) error {
+// Upload stores a number of uploadable items.
+// It first queries the CAS to see which items are missing and only uploads those that are.
+func (c *Client) Upload(ctx context.Context, data ...*chunker.Chunker) error {
```
Comment (ramymedhat) [inline go/client/cas.go]: suggestion: UploadIfMissing?

File: go/client/cas_test.go [hunk 2]
```diff
@@ -390,34 +376,42 @@ func TestWriteBlobs(t *testing.T) {
...
 					for _, blob := range tc.input {
-						input[digest.NewFromBlob(blob)] = blob
+						input = append(input, chunker.NewFromBlob(blob, 20))
 					}
 
-					err := c.WriteBlobs(ctx, input)
+					err := c.Upload(ctx, input...)
 					if err != nil {
-						t.Fatalf("c.WriteBlobs(ctx, inputs) gave error %s, expected nil", err)
+						t.Fatalf("c.Upload(ctx, input) gave error %v, expected nil", err)
```
Comment (ramymedhat) [inline go/client/cas_test.go]: Errorf

File: go/client/cas_test.go [hunk 3]
```diff
@@ -390,34 +376,42 @@ func TestWriteBlobs(t *testing.T) {
...
 
-					for _, blob := range tc.present {
-						dg := digest.NewFromBlob(blob)
-						if fake.BlobWrites(dg) > 0 {
-							t.Errorf("blob %v with digest %s was uploaded even though it was already present in the CAS", blob, dg)
+					for _, ch := range input {
+						dg := ch.Digest()
+						blob, err := ch.FullData()
+						if err != nil {
+							t.Fatalf("ch.FullData() returned an error: %v", err)
```
Comment (ramymedhat) [inline go/client/cas_test.go]: Errorf

File: go/client/cas_test.go [hunk 4]
```diff
@@ -390,34 +376,42 @@ func TestWriteBlobs(t *testing.T) {
...
+						}
+						if present[dg] {
+							if fake.BlobWrites(dg) > 0 {
+								t.Errorf("blob %v with digest %s was uploaded even though it was already present in the CAS", blob, dg)
+							}
+							continue
```
Comment (ramymedhat) [inline go/client/cas_test.go]: How do you make sure the blobs where actually uploaded and not only that the chunkers pulled all the data?

File: go/client/cas.go (hunk 1 above)
Comment (ola-rozenfeld) [reply go/client/cas.go]: Sure.

File: go/client/cas_test.go (hunk 2 above)
Comment (ola-rozenfeld) [reply go/client/cas_test.go]: Sorry, I keep forgetting :-)

File: go/client/cas_test.go (hunk 4 above)
Comment (ola-rozenfeld) [reply go/client/cas_test.go]: fake.BlobWrites returns how many times Write or Batch upload was called on the digest. I don't actually know that the chunkers pulled all the data, but I verify in the next lines that the data is correct, so the conclusion is that it was uploaded.

Comment (ramymedhat) [reply go/client/cas_test.go]: sgtm

**Pull request merged**